Add optional start_value to generateSerialID

raimannma opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes into CHANGELOG.md):
Added optional `start_value` parameter to `generateSerialID` function to specify custom starting values for new series.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

**Motivation:** 
Allows users to specify a custom starting value for `generateSerialID` counters instead of always starting at 0.

**Parameters:**
- `start_value` (UInt64, optional): Initial counter value for new series. Defaults to 0. Only affects series creation; subsequent calls continue incrementing.

**Example use:**
```sql
-- Start a new series at 100
SELECT generateSerialID('customer_id', 100);  -- Returns: 100
SELECT generateSerialID('customer_id', 100);  -- Returns: 101

-- Generate multiple IDs
SELECT generateSerialID('order_id', 1000) FROM numbers(3);
-- Returns: 1000, 1001, 1002
```

raimannma commented:
@antaljanosbenjamin Do I have to do anything to get the `can_be_tested ` label?

antaljanosbenjamin commented:
Ahh sorry, I wanted to add it, but messed up somehow.

raimannma commented:
`python3 -m ci.praktika run "Fast test" --test 03129_serial_test_zookeeper`

Test passes locally now, you know what's wrong with the Pipeline @antaljanosbenjamin 
Seems like the workflow is missing permissions?! Maybe you have to trigger it?

antaljanosbenjamin commented:
Now it should run properly, let's see the results.

antaljanosbenjamin commented on src/Functions/generateSerialID.cpp:
```diff
@@ -88,14 +90,19 @@ class FunctionSerial : public IFunction
-        validateFunctionArguments(*this, arguments, mandatory_args);
+
+        FunctionArgumentDescriptors optional_args{
+            {"start_value", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), &isColumnConst, "const UInt"}
```
When the argument is not bounded to a specific size of integer, we indicate that with a `*`.
```suggestion
            {"start_value", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), &isColumnConst, "const UInt*"}
```

antaljanosbenjamin commented on src/Functions/generateSerialID.cpp:
```diff
@@ -110,6 +117,11 @@ class FunctionSerial : public IFunction
             throw Exception(ErrorCodes::BAD_ARGUMENTS, "Series name '{}' is too long", series_name);
 
+        UInt64 start_value = 0;
+        bool has_start_value = arguments.size() >= 2;
```
```suggestion
        const auto has_start_value = arguments.size() >= 2;
```

antaljanosbenjamin commented on src/Functions/generateSerialID.cpp:
```diff
@@ -126,7 +138,10 @@ class FunctionSerial : public IFunction
         else
-            keeper->createIfNotExists(serial_path, "0");
+        {
+            String initial_value = has_start_value ? toString(start_value) : "0";
```
I think it worth the readability.
```suggestion
            const auto initial_value = toString(start_value);
```

antaljanosbenjamin commented on src/Functions/generateSerialID.cpp:
```diff
@@ -164,14 +179,15 @@ REGISTER_FUNCTION(Serial)
         .description=R"(
 Generates and returns sequential numbers starting from the previous counter value.
-This function takes a constant string argument - a series identifier.
+This function takes a constant string argument - a series identifier, and an optional starting value.
```
Let's mention that the start value is ignored if the counter already exist, similarly to how you added the example.

raimannma replied:
Yeah :+1: I would add that to the argument description.

antaljanosbenjamin commented:
> Is the failing test something I need to fix? I saw it passed in other CIs but also failed in some PRs

It is not caused by your PR in my opinion, so I just restarted the CI.

raimannma committed:
feat(Functions): Add optional start_value to generateSerialID

This commit introduces the ability to specify a starting value for a new series in the generateSerialID function.
